Add development guidelines (CONTRIBUTING.md)

Technici4n opened the pull request:
**[Rendered View](https://github.com/Technici4n/fabric/blob/development-guidelines/CONTRIBUTING.md)**

This PR adds a `CONTRIBUTING.md` document encoding Fabric API development guidelines. This will ensure that everyone can check what our contribution standards are. The goal is to ultimately make writing and reviewing Fabric PRs a better experience.

This document has been in the works for a long time. I want to give a special thanks to @i509VCB and @Vaerian for their important contributions to the draft that was made over a year ago, and the rest of team who provided valuable feedback.

I would like to ask those who are interested in contributing to Fabric API and have some time to read this document, and comment on this PR if something is not clear.

Note that there are still a few TODO items, I will be addressing them over the coming days.

Technici4n committed:
Add development guidelines (CONTRIBUTING.md)

modmuss50 reviewed (approved):
Great start, this is something we can contiune to add to and adapt as required. 👍

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+    - Remember that deprecating code is always an option.
+
+### Targeted Minecraft versions
+- Fabric API does not have strict Minecraft version support policies, but rathers supports what is feasible and what the community is interested in.
```
```suggestion
- Fabric API does not have strict Minecraft version support policies, but rather supports what is feasible and what the community is interested in.
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+- Builders can be used instead of constructors or factory methods with large amounts of parameters.
+    - TODO: any specific builder guidelines?
+
+### API class modifiers and member visiblity
```
```suggestion
### API class modifiers and member visibility
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+
+### Mixins
+
+These guidelines should be followed with regards to a mixin's visiblity and naming:
```
```suggestion
These guidelines should be followed with regards to a mixin's visibility and naming:
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+| | Accessors | Mixin |
+| -------- | -------- | -------- |
+| Naming   | **Target**Accessor | **Target**Mixin (May include `Client/Server` or `Legacy` prefix if needed) |
+| Visiblity | public | package-private* |
```
```suggestion
| Visibility | public | package-private* |
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+    - Complex class hierarchies.
+- Functional operators (i.e. interfaces) should be designed for maximal readability, and only used when necessary.
+    - Ask for concrete objects if applicable.
+    - Don't include superflous methods, for example replace shouldApply+apply with a single apply that returns success.
```
```suggestion
    - Don't include superfluous methods, for example replace shouldApply+apply with a single apply that returns success.
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+
+Since Minecraft has two primary threads, the render and server threads, APIs dealing with shared state need to consider thread safety. This will usually require consideration on a case-by-case basis.
+
+- A cache shared between the render and the server threads may for example wish to use locking for unfrequent accesses, or otherwise a copy-on-write strategy to ensure lock-free reads and thread-safe writes.
```
```suggestion
- A cache shared between the render and the server threads may for example wish to use locking for infrequent accesses, or otherwise a copy-on-write strategy to ensure lock-free reads and thread-safe writes.
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+
+Every module should in its `fabric.mod.json` declare dependencies for:
+- `"fabricloader": ">=x.y.z"`, where `x.y.z` is the version used at the time the module is added.
+- Other used modules, for example `"fabric-api-base": "*"` if events are used. No explicit version needs to be specificed.
```
```suggestion
- Other used modules, for example `"fabric-api-base": "*"` if events are used. No explicit version needs to be specified.
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+The organization of mixins with a package is dependent on the type of module.
+- Generally if there are a small amount of mixins, then having all mixins in the same package is fine.
+- Client only and dedicated server only mixins should be moved into `client/server` subpackages.
+- If a module is generally complex or has multiple distinct parts, multiple mixins for each class target may exist assuming said mixins are seperated into subpackages by feature.
```
```suggestion
- If a module is generally complex or has multiple distinct parts, multiple mixins for each class target may exist assuming said mixins are separated into subpackages by feature.
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+
+### Style guidelines
+
+These are less strict guildelines for the code style, but you should generally obey these:
```
```suggestion
These are less strict guidelines for the code style, but you should generally obey these:
```

apple502j commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,428 @@
+- Good documentation doesn't only explain what something is, but (as appropriate) why it exists, what are the intended use cases, what use cases something is not suitable for and any semantics that need to be kept in mind.
+    - In particular, tricky implementation details should be explained by a few comments when appropriate.
+- Direct references to classes or members should use `@link` tag, unspecific further reading elsewhere may use `@see`.
+- Deprecated elements should describe and reference the replacement or alternative in Javadoc ``@deprecated`. The ``@deprecated` tag has to be last, everything else may be removed while adding it.
```
```suggestion
- Deprecated elements should describe and reference the replacement or alternative in Javadoc `@deprecated`. The `@deprecated` tag has to be last, everything else may be removed while adding it.
```

Technici4n committed:
Fix typos

Co-authored-by: apple502j <33279053+apple502j@users.noreply.github.com>

Technici4n committed:
Address remaining TODOs, a few formatting changes

liach commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+
+The rest of this document is split in the following categories:
+
+- **General design considerations**: Broad guidelines to keep in mind when writing APIs for Fabric.
```
```suggestion
- [**General design considerations**](#general-design-considerations): Broad guidelines to keep in mind when writing APIs for Fabric.
```
we can add such links, same for other categories below,

liach commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+Since Minecraft has two primary threads, the render and server threads, APIs dealing with shared state need to consider thread safety. This will usually require consideration on a case-by-case basis.
+
+- A cache shared between the render and the server threads may for example wish to use locking for infrequent accesses, or otherwise a copy-on-write strategy to ensure lock-free reads and thread-safe writes.
+- Read-only access from multiple threads does not need synchronization.
```
I hope you add another entry for lazy initialization (like benign race), such as:
```java
private Stuff cachedStuff;

private Stuff stuff() {
  var stuff = this.cachedStuff;
  if (stuff != null) return stuff;
  return this.cachedStuff = compute();
}
```

Frequently, people will write code like
```java
private Stuff stuff() {
  if (this.cachedStuff != null) return this.cachedStuff;
  return this.cachedStuff = compute();
}
```
which reads `cachedStuff` field twice and is subject to races, as the first read may be non-null but the second one may.

Technici4n replied:
Both are correct provided that `cachedStuff` is `volatile`. The reason we use an intermediate variable is to avoid two `volatile` reads, which are more expensive than a single `volatile` read. I'll add something in that area.

apple502j replied:
Are there actually APIs where this matters, btw?

liach replied:
This is a point of concern in the API implementation, especially that there are server and client threads

apple502j replied:
I don't see APIs with this use pattern intended to be called on both sides, but guess there can be.

Technici4n replied:
You can find it in the depths of the transfer API: https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/ItemVariantImpl.java#L144-L153

It's also used in the content registries:
https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/FlammableBlockRegistryImpl.java#L49-L70
https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/FuelRegistryImpl.java#L51-L59

liach commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+    - Any member, parameter or return value that is not marked as nullable can be assumed to be nonnull. The `@NotNull` annotation should never be used, it is implicit.
+    - Introducing custom types might be appropriate (no need to specify `List<@Nullable String>` everywhere).
+- Deprecated API members should have the `@Deprecated` annotation.
+    - Avoid specifying `forRemoval = true`, the functionality will be supported as long as possible.
```
Not that we usually intend to support such deprecated APIs; the true intention is that we usually sunset an API with a whole module, and we only use `forRemoval = true` if we know this API must be removed as the current module lives and before we move to a new module, which is exceedingly rare.

Technici4n replied:
I think that this part is fine as-is: `Avoid specifying` means that in some edge cases we can still allow it (vs `Never specify` which would always forbid it).

liach commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+### Deprecated modules
+Modules that are entirely deprecated are in the `deprecated/` folder. All of their API classes should be deprecated, and their `fabric.mod.json` should have the deprecated lifecycle marker. (See PR checklist below).
+
+A module may only be deprecated if all of its functionality is also provided by a non-deprecated module.
```
... or those non-migrated functionalities are determined to be no longer feasible

Juuxel replied:
Including "by a non-deprecated module **or vanilla Minecraft itself**" might be good, some loot tables v1 api was later added to the game and dropped from v2.

Technici4n replied:
I added the note about vanilla MC.

> ... or those non-migrated functionalities are determined to be no longer feasible

I don't think this is fully accurate, as we should try to provide a reasonable alternative. After all, if the functionality was added in the first place, it was probably for a good reason. So I suggest not to add this, and we can discuss exceptions on a case-by-case basis.

liach commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+Modules whose design is hard to evaluate might go through an experimental phase, allowing for relaxed backwards compatibility requirements, as described in the Backwards compatibility section.
+
+Writers of experimental modules need to consider the following additional requirements:
+- All API classes should be marked as `@ApiStatus.Experimental`. Do not use `@Deprecated` for experimental modules.
```
```suggestion
- All API classes should be marked as `@ApiStatus.Experimental`. Do not use `@Deprecated` for experimental modules to generate compiler warnings.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- Transitive Access Wideners (TAWs) should be used to expose access to private or protected members in vanilla classes.
+    - Most TAWs should go in the dedicated `fabric-transitive-access-wideners-v1` module.
+        - Remember to add the `transitive-` prefix, otherwise dependent mods will not see the access modifications.
+        - Some TAWs such as Block subclass constructors are automatically generated. Make sure that you don't edit the `.accesswidener` file directly. Rather edit the template file (`.accesswidener.template`) and run the `gradlew generateAccessWidener` task to update the generated file.
```
```suggestion
        - Some TAWs such as Block subclass constructors are automatically generated. Make sure that you don't edit the `.accesswidener` file in `src` directly. Rather edit the template file (`template.accesswidener`) and run the `gradlew generateAccessWidener` task to update the generated file.
```
See the file name here: https://github.com/FabricMC/fabric/blob/1.19/fabric-transitive-access-wideners-v1/template.accesswidener

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+        - Large amounts of TAWs for a specific purpose can be included in another module, as is the case for the data generation API, for example.
+    - Do **not** expose TAWs for functions that take a `String` identifier.
+        - This makes it too easy to forget the mod ID namespace, so the identifier would often end up in the vanilla `minecraft` namespace.
+        - In general, keep the API guidelines in mind when deciding whether something should be a TAW.
```
```suggestion
      - In general, keep the API guidelines in mind when deciding whether something should be a TAW.
```
This isn't really specific to string IDs

Juuxel commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,423 @@
+        default void injectedMethod() {
+		    throw new UnsupportedOperationException("Implemented via mixin");
+        }
+        ```
````
Fixed funky indentation on the middle line:
```suggestion
        ```java
        default void injectedMethod() {
        	throw new UnsupportedOperationException("Implemented via mixin");
        }
        ```
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- Limited duplication can be better than indirection, unless the code is complex or used several times.
+- Indirections might make the code harder to read, and should be weighed against their benefits. Examples include:
+    - Lambda composition (forEach, streams).
+    - Method calls.
```
Method calls are perfectly normal Java code and not particularly hard to read?

Technici4n replied:
The intent is already conveyed by `Method splitting`, so I'm removing this one.

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+    - Complex class hierarchies.
+- Functional operators (i.e. interfaces) should be designed for maximal readability, and only used when necessary.
+    - Ask for concrete objects if applicable.
+    - Don't include superfluous methods, for example replace shouldApply+apply with a single apply that returns success.
```
```suggestion
    - Don't include superfluous methods. For example, replace `shouldApply`+`apply` with a single `apply` that returns success.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- Every API class should carry a Javadoc block explaining its purpose and reference related classes. Example code in the primary class of any major feature should outline the use, including related vanilla invocations/registrations/etc as applicable to provide an idea of how to start.
+    - Parameters, implementation bodies, etc... may be omitted as appropriate, since the examples are not meant as to be fully working implementations.
+    - The examples should be more akin to pseudocode with a checklist for the process and pointers to everything needed.
+    - These examples should be written for developers that have a knowledge of java and basic knowledge of the Minecraft codebase.
```
```suggestion
    - These examples should be written for developers that have a knowledge of Java and basic knowledge of the Minecraft codebase.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+    - Modules whose primary purpose is not interaction with their API do not need this suffix. For example, `fabric-transitive-access-wideners-v1` or `fabric-convention-tags-v1`.
+- Module names should always be suffixed by a major version (`-v1`, `-v2`, etc).
+    - The major version starts at `v1` for new functionality, unless they replace a module with equivalent functionality, in which case the version is incremented.
+    - `vn` and `vn+1` module names need not match exactly. For example, `fabric-loot-table-api-v1` was replaced by `fabric-loot-api-v2`.
```
```suggestion
    - `vn` and `vn+1` module names need not match exactly. For example, `fabric-loot-tables-v1` was replaced by `fabric-loot-api-v2`.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- Statements should use the block form except for single-line if statements that use a short condition and nothing else.
+- Ternary operator use only within a single line total.
+
+Fabric API uses the Gradle checkstyle plugin. If you run `gradle check` any style errors will be logged in the output. The style errors can also be viewed in a generated webpage by the gradle task.
```
```suggestion
Fabric API uses the Gradle checkstyle plugin. If you run `gradle check` any style errors will be logged in the output. The style errors can also be viewed in a generated webpage by the Gradle task.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+## Checklist before submitting a pull request
+
+### Apply license headers
+There is a gradle task that can automate this for you. Simply run `gradlew spotlessApply`.
```
```suggestion
There is a Gradle task that can automate this for you. Simply run `gradlew spotlessApply`.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+Fabric API pull requests should be tested in the dev environment and in production (on both a client and dedicated server).
+The `gradlew build` command can be used to produce the Fabric API fatjar, located in `builds/libs/`.
+
+#### Common Mistakes
```
```suggestion
#### Common mistakes
```

IMB11 replied:
Actually, you're incorrect.

According to Oxford:

> Capitalise the first word of the title, and all words within the title except articles (a/ an/the), prepositions (to/on/for etc) and conjunctions (but/and/or etc).

Technici4n replied:
The rest of the document is written with only the first words of the titles capitalized. It looks better like this than with most words capitalized. ;)

IMB11 replied:
Fair

Juuxel replied:
Yep, I suggested this for consistency with the rest of the document (I personally couldn't care less about which style is used as long as they're not mixed, which looks ugly)

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- `@ApiStatus.Internal` is recommended for implementation classes to avoid IDE autocompletion suggesting them. There should be no need for internal methods or classes in the public API package.
+
+### Naming
+- All names should follow the yarn naming standards.
```
```suggestion
- All names should follow the [Yarn naming standards](https://github.com/FabricMC/yarn/blob/HEAD/CONVENTIONS.md).
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,423 @@
+- Simple code that is easy to debug and reason about is generally preferable to the shortest possible implementation.
+- Limited duplication can be better than indirection, unless the code is complex or used several times.
+- Indirections might make the code harder to read, and should be weighed against their benefits. Examples include:
+    - Lambda composition (forEach, streams).
```
Nitpick: `forEach` on `Iterable`s doesn't really compose anything, not does it make anything harder to read. I still agree it should be avoided for other reasons though (no `break`/`continue`, possibly slower in hot code)

Technici4n replied:
Replaced with `Lambda methods`.

Technici4n committed:
Apply feedback

- Add links to sections in summary
- Add note about thread-safe caching
- Functionalities added by vanilla MC don't need to be ported to new module versions
- Do not use deprecated to generate compiler/IDE warnings
- TAW updates
- Interface injection indents
- A few more fixes

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>
Co-authored-by: liach <7806504+liach@users.noreply.github.com>

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+    - This pattern may be used if the backing implementation may be changed by another mod (such as the renderer api). Even then there should be static methods to access the standard api facilities.
+- Avoid unnecessary use of generics.
+    - Unless Vanilla Minecraft mandates the use of generics, or a good reason exists to use generics.
+    - In particular, use of self types should be avoided.
```
Why?

Technici4n replied:
Well, they lead to many problems and should be avoided for that reason.

Juuxel replied:
> many problems

...that doesn't really answer anything.

Technici4n replied:
I don't have a specific example in mind, but I had issues when using them in the past in AE2. `IAEStack<T extends IAEStack<T>>` used to be generic hell.

Juuxel replied:
Self types are fine - for example vanilla's loot builders use them without them being generic hell. Using generics (including self types) is one thing, going overboard with them is another.

Technici4n replied:
I removed this line since it's already contained in `Avoid unnecessary use of generics.`

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+    - In general, keep the API guidelines in mind when deciding whether something should be a TAW.
+- Interface injection (i.e. making a minecraft class or interface extend a Fabric interface) should be considered over separate static helpers.
+    - Interface injection requires both a `fabric.mod.json` custom value to make it visible in Minecraft source code, and a mixin to actually implement the interface at runtime.
+    - Injected interfaces should have **no abstract methods**.
```
Might be good to include the rationale here (that javac will throw a fit if it can't find the implementation of an itf method on a class iirc)

Technici4n replied:
Well, obviously we don't want to force modders to add methods beyond what vanilla requires. :P

Juuxel replied:
Well yeah, that too. But the javac thing applies when you're injecting into a final class or a class that no one will ever want to extend (LootTable)

Technici4n replied:
Should be ok now.

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+- Classes in Fabric API should be `final` classes unless the class exposed in the API is explicitly meant for extension.
+- `private` constructors should be used in API classes unless the class is explicitly meant to be instantiated.
+    - This only applies to modder-facing classes, i.e. classes in the `net.fabricmc.fabric.api` subpackage. See below for package structure.
+    - They should be placed at the very bottom of the class to not hurt readability.
```
Placing them at the very top also doesn't hurt readability imo (JDK code does this)

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+### Events
+
+- Events should not be used if there is only one subscriber, like a handler in a registered unique namespace.
+- Events should be produced and fully usable without object allocation.
```
```suggestion
- Events should be produced and fully usable without creating data holder objects for inputs.
```
Events sometimes do need to allocate objects, for example when they use `TypedActionResult`

Technici4n replied:
It's a general design guideline. While it mainly applies for inputs, it could also apply for outputs in some cases (e.g. pass a `Consumer` parameter instead of returning a `List`).

Juuxel replied:
The current version reads to me like "you should never allocate objects when designing events", which does not sound like a soft general guideline

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+- Events should use dedicated callback interfaces.
+    - Callback interfaces should be `@FunctionalInterface`s.
+    - Callback methods should be uniquely named such that a handler can implement multiple at once.
+        - Avoid words that are already clear from the parameters. For example, prefer `onStartTick(MinecraftServer)` over `onStartServerTick(MinecraftServer)`. The interface should still be named `StartServerTick`.
```
```suggestion
        - Avoid words that are already clear from the parameters. For example, prefer `onStartTick(MinecraftServer)` over `onStartServerTick(MinecraftServer)`. The interface should still be named `StartServerTick` if "server" is not already implied by the containing class.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+
+#### Event naming
+
+- Callback interfaces.
```
```suggestion
- Callback interfaces:
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+    - Method names for notification events should be prefixed with `on`.
+      For example, a `DataLoad` event would have an `onDataLoad` method.
+    - Events that may allow or block some action should start with `Allow`. For example, an event to cancel player death might be called `AllowPlayerDeath`, with method name `allowDeath`.
+- `Event<>` fields and methods.
```
```suggestion
- `Event<>` fields and methods:
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+- Simple code that is easy to debug and reason about is generally preferable to the shortest possible implementation.
+- Limited duplication can be better than indirection, unless the code is complex or used several times.
+- Indirections might make the code harder to read, and should be weighed against their benefits. Examples include:
+    - Lambda methods (forEach, streams).
```
```suggestion
    - Lambda methods (`forEach`, streams).
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+
+## Documentation
+
+- Every API class should carry a Javadoc block explaining its purpose and reference related classes. Example code in the primary class of any major feature should outline the use, including related vanilla invocations/registrations/etc as applicable to provide an idea of how to start.
```
```suggestion
- Every API class should carry a Javadoc comment explaining its purpose and reference related classes. Example code in the primary class of any major feature should outline the use, including related vanilla invocations/registrations/etc as applicable to provide an idea of how to start.
```

Juuxel commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,437 @@
+- A brief description belongs in the 1st paragraph, with further paragraphs separated by blank lines and starting with `<p>`. Javadoc for methods may follow with another blank line before describing parameters, return values and exceptions with the appropriate tags. All accessible members should be described appropriately.
+- Good documentation doesn't only explain what something is, but (as appropriate) why it exists, what are the intended use cases, what use cases something is not suitable for and any semantics that need to be kept in mind.
+    - In particular, tricky implementation details should be explained by a few comments when appropriate.
+- Direct references to classes or members should use `@link` tag, unspecific further reading elsewhere may use `@see`.
```
```suggestion
- Direct references to classes or members should use the `@link` and `@linkplain` tags, unspecific further reading elsewhere may use `@see`.
```

Juuxel reviewed (commented):
This should be the last round of comments from me :tiny_potato:

Technici4n committed:
Apply suggestions from code review

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

Technici4n committed:
Apply suggestions from Juuz

Technici4n committed:
Remove mention of self types

deirn commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,439 @@
+
+### Run checks
+- The `gradlew check` task runs all the style checks and game tests.
+- Code style can be checked with the faster `gradlew checkstyleMain checkstyleTestmod`.
```
Maybe create a new task that run all the checkstyle tasks?
```gradle
task checkstyle {
	dependsOn tasks.withType(Checkstyle)
	subprojects.forEach {
		dependsOn it.tasks.withType(Checkstyle)
	}
}
```

Technici4n replied:
@modmuss50 what do you think?

modmuss50 replied:
Yeah, something along those lines sounds reasonable.
